archunit/src/test/java/com/tngtech/archunit/testutil/assertion/ConditionEventsAssertion.java

```diff
@@ -0,0 +1,99 @@
+package com.tngtech.archunit.testutil.assertion;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import com.google.common.collect.ImmutableSet;
+import com.tngtech.archunit.lang.ConditionEvent;
+import com.tngtech.archunit.lang.ConditionEvents;
+import org.assertj.core.api.AbstractCharSequenceAssert;
+import org.assertj.core.api.AbstractIterableAssert;
+import org.assertj.core.api.Assertions;
+import org.assertj.core.api.ObjectAssert;
+import org.assertj.core.api.ObjectAssertFactory;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Lists.newArrayList;
+
+public class ConditionEventsAssertion
+        extends AbstractIterableAssert<ConditionEventsAssertion, ConditionEvents, ConditionEvent, ObjectAssert<ConditionEvent>> {
+
+    public ConditionEventsAssertion(ConditionEvents actual) {
+        super(actual, ConditionEventsAssertion.class);
+    }
+
+    public void containViolations(String violation, String... additional) {
+        Assertions.assertThat(actual.containViolation()).as("Condition is violated").isTrue();
+
+        List<String> expected = concat(violation, additional);
+        if (!sorted(messagesOf(actual.getViolating())).equals(sorted(expected))) {
+            failWithMessage("Expected %s to contain only violations %s", actual, expected);
+        }
+    }
+
+    public void containAllowed(String message, String... additional) {
+        Assertions.assertThat(actual.getAllowed().size()).as("Allowed events occurred").isGreaterThan(0);
+
+        List<String> expected = concat(message, additional);
+        if (!sorted(messagesOf(actual.getAllowed())).equals(sorted(expected))) {
+            failWithMessage("Expected %s to contain only allowed events %s", actual, expected);
+        }
+    }
+
+    private List<String> messagesOf(Collection<? extends ConditionEvent> events) {
+        final List<String> result = new ArrayList<>();
+        for (ConditionEvent event : events) {
+            result.addAll(event.getDescriptionLines());
+        }
+        return result;
+    }
+
+    private List<String> concat(String violation, String[] additional) {
+        ArrayList<String> list = newArrayList(additional);
+        list.add(0, violation);
+        return list;
+    }
+
+    private List<String> sorted(Collection<String> collection) {
+        ArrayList<String> result = new ArrayList<>(collection);
+        Collections.sort(result);
+        return result;
+    }
+
+    public void containNoViolation() {
+        Assertions.assertThat(actual.containViolation()).as("Condition is violated").isFalse();
+        Assertions.assertThat(messagesOf(actual.getViolating())).as("Violating messages").isEmpty();
```

Comment:
Can this ever trigger if `actual.containViolation()` is already known to be `false` at this point?

Reply:
At the moment not I guess, but I thought that is implementation detail. Yes, it should always be in sync, but I would want to know if it is not? Or do you think this is the wrong place for it?

Reply:
I just was sceptical about – currently – dead code, but I believe that it doesn't hurt.
(It should in fact even work if it once gets executed... 😉)

Reply:
I was thinking more about maintenance in the future. If this should not hold anymore at some point I would want to know. Or in other words, my test fixture is, that these assumptions both hold always, I don't think it's necessary that the current implementation would allow the second assumption to fail individually for that. It's more for the future :wink: